### DISCUSSION TITLE
How does one use getTableItem (TS SDK) against a Table with inner vector data?

### DISCUSSION THREAD

**Discussion by Alexzander-Stone** [aptos-labs/aptos-developer-discussions#157] [Questions] [answered]
### Describe your question in detail.

Asked by `Kushagra` on Discord (#dev-discussions):
>Hi, I have a create a table in my move module
>```
>struct ReportsObject has key {
>        reports: Table<String, vector<Asset>>,
>}
>```
>I have invoked mu module function and inserted a asset in the table. I have the table handle now.
>
>Now I am trying to fetch table items using 
>```
>let resources = await aptos.getTableItem({
>            handle: <handle>,
>            data: {
>                key: 'key',
>                key_type: 'string',
>                value_type: '0x508c3346b5cf6e78a2053da32d2d3529a0758816ba95e97e01b334a1398eff9a::<name>::Asset'
>            },
>        });
>```

Where they then ask how to supply MoveType's to getTableItem.

### What error, if any, are you getting?

>I am getting the following error
>`AptosApiError: Fullnode error: {"message":"'table_item_request' invalid: Unable to parse move type string","error_code":"invalid_input","vm_error_code":null}`
>

### What have you tried or looked at?  Or how can we reproduce the error?

>I have reffered to https://aptos-labs.github.io/ts-sdk-doc/types/Types.MoveType.html
>but still unable to find a valid value to pass for key_type

### Which operating system are you using?

N/A

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by Alexzander-Stone** [comment, marked as answer]
### Two Problems Identified
#### 1. key_type
For `key_type` in `getTableItem`, you'll need to provide the full struct reference, as defined by [MoveType](https://aptos-labs.github.io/ts-sdk-doc/types/Types.MoveType.html#:~:text=non%2Dreference%20MoveTypeId%7D%3E-,struct,-%3A%20%7Baddress%7D%3A%3A%7Bmodule_name).

A reminder, String is a struct, not a primitive type (e.g. [integers](https://aptos.dev/move/book/integers/) and others).

#### 2. value_type
`value_type` is also misrepresented, since your table contains a `vector` of `Asset`, and not just a single `Asset`.


### Solutions

#### 1. key_type
In your example, it would require changes from:
```
key_type: 'string'
```
to:
```
key_type: '0x1::string::String'
```

I derived this struct reference from the [Move reference documentation](https://aptos.dev/reference/move/?branch=mainnet&page=move-stdlib/doc/string.md) on Move-stdlib's string.

#### 2. value_type
You'll need to include the vector as part of the string, resulting in changes from:
```
value_type: '0x508c3346b5cf6e78a2053da32d2d3529a0758816ba95e97e01b334a1398eff9a::<name>::Asset'
```
to:
```
value_type: 'vector<0x508c3346b5cf6e78a2053da32d2d3529a0758816ba95e97e01b334a1398eff9a::<name>::Asset>'
```

------


### Useful links
* [Aptos-core example](https://github.com/aptos-labs/aptos-core/blob/d8c0c7ffb8e33ee9c83c13584126985d52cdaa3d/ecosystem/typescript/sdk/src/providers/aptos_client.ts#L899) of using getTableItem
* [Testnet demonstration](https://explorer.aptoslabs.com/account/0x44616c94a2326734bb0d03dd5dc26124f1c041462cc0fccd241b510c414618ca/modules/code/get_table_item_example?network=testnet) I deployed
* [Helper class](https://github.com/aptos-labs/aptos-ts-sdk/blob/f8101007dbd0e59d7aa6954b5031d169f178d483/tests/e2e/helper.ts) used to create the aptos object

Additionally, here's the snippet of Typescript code to retrieve the Table entry from the Testnet demonstration:
```
let resources = await aptos.getTableItem({
    handle: "0x90b53c1ac994d85235d112f43ff66809af80cf878592a38dacd2442da3b09a77",
    data: {
        key: 'this_is_a_key',
        key_type: '0x1::string::String',
        value_type: 'vector<0x44616c94a2326734bb0d03dd5dc26124f1c041462cc0fccd241b510c414618ca::get_table_item_example::Asset>'
    },
});
```
which results in:
`[ { value: '803' } ]`

**Reply by deepkushagra** [reply]
Thanks for the quick resolution. This solution is working.

**Reply by deepkushagra** [reply]
Hi @Alexzander-Stone 
I have converted my javascript object to MoveString to pass it in transactions
```
const moveString2 = new MoveString({name: 'name', add: 'address'});
moveString2.serializeForEntryFunction(new Serializer());
```
After fetching this value back from the move module on devnet, I want to convert it back to javascript object.
How should we do it? Is there a reference ?

**Reply by Alexzander-Stone** [reply]
Hey there, @deepkushagra.

Are you asking how to convert `moveString2` into a JS object or the return value of `aptos.getTableItem(...)`? Or is it something else, entirely?

Also, do you mind sharing a code snippet of the initialization/declaration of the item that you're hoping to convert? It'll help, a ton.

**Reply by deepkushagra** [reply]
After serialising the string I passed it in for transactions
```
const moveString2 = new MoveString({name: 'name', add: 'address'});
moveString2.serializeForEntryFunction(new Serializer());
const transaction = await aptos.transaction.build.simple({
            sender: accountAddress,
            data: {
                function: `${accountAddress}::<name>::create_data`,
                functionArguments: [moveString, moveString1, moveString2],
            },
        });
        const committedTransaction = await aptos.signAndSubmitTransaction({
            signer: signerArg,
            transaction,
        });
```
However, after fetching the the string back from the table.
I have that string which I submitted in the transaction before. I am unable to convert the returned value to a javascript object as before
`{name: 'name', add: 'address'}`
This was my object initially

**Reply by Alexzander-Stone** [reply]
So if I'm understanding correctly, @deepkushagra, you're:
1. Creating a MoveString with a JSON object as the content.
2. Serializing the MoveString into bytes to pass into the entry function.
3. Submitting a TX where the previously mentioned MoveString is stored in a Table on-chain.
4. Call aptos.getTableItem(...) against the Table to grab the MoveString that was just stored.
5. Attempt to convert the MoveString back into a JSON object.

But, after step 4, you're not able to convert it back into a JSON object. 

If my understanding is complete, can you share a code snippet for steps 4 and 5?

**Reply by deepkushagra** [reply]
Yes, you understood it right @Alexzander-Stone 
My code for step 4-5
```
let resources4 = await aptos.getTableItem({
            handle: resources2.reports?.handle,
            data: {
                key: 'KEY',
                key_type: '0x1::string::String',
                value_type: 'vector<0x508c3346b5ajd8898a2053da32d2d3529a0758816ba95e97e01b334a1398ensjsa::<name>::Asset>'
            },
        });
        let asset = resources4[0].asset;
```
asset is printed in console as:-  `asset: '[object Object]'`
I have tried various ways like, parsing it using JSON.parse(). Making a MoveString() from it.
What would be the correct way?

**Reply by Alexzander-Stone** [reply]
Have you tried using `JSON.stringify` on the object, before creating a MoveString out of it? 

Something like:
```
const moveString2 = new MoveString(JSON.stringify({name: 'name', add: 'address'}));
...
```


It appears that MoveString won't automatically `stringify` the object, resulting in `[object Object]`. Performing it yourself should fix the problem.
